diffkemp/simpll/InstPatternComparator.cpp

```diff
@@ -0,0 +1,660 @@
+//===---- InstPatternComparator.cpp - Code pattern instruction matcher ----===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of the LLVM code pattern matcher. The
+/// pattern matcher is a comparator extension of the LLVM FunctionComparator
+/// tailored to comparison of general instruction-based patterns.
+///
+//===----------------------------------------------------------------------===//
+
+#include "InstPatternComparator.h"
+#include "Config.h"
+
+/// Compare the module function and the difference pattern from the starting
+/// module instruction. This includes checks for correct input mappings.
+int InstPatternComparator::compare() {
+    // Clear all previous results.
+    beginCompare();
+    InstMatchMap.clear();
+    PatMatchMap.clear();
+    ModMatchMap.clear();
+
+    // Run the main matching algorithm.
+    if (int Res = matchPattern())
+        return Res;
+
+    // Reset the comparison state without clearing pattern matches.
+    beginCompare();
+
+    // Ensure that the created input mapping is correct. All input instructions
+    // and arguments have to be mapped correctly.
+    if (int Res = checkInputMapping())
+        return Res;
+
+    return 0;
+}
+
+/// Compare a module input value with a pattern input value. Used for comparing
+/// input values that could not be mapped during the first one-side comparison.
```

Comment:
This function tries to confirm that the input value of a pattern instruction can be matched to a module counterpart. This is important since we start matching a pattern somewhere in the middle of the corresponding module, so the values used by some instructions do not get compared against each other right away.

I will try to think of an example, since I do agree that this might be the least understandable concept of all. Also, this raises the question of whether doing this at all is necessary, but IIRC we have already agreed that it is.

Reply:
Alright, I think I have found  good example. It's based on the `rhel-82-83-patterns_blk_execute_rq` test case.

As stated above, since we (generally) match patterns in the middle of module functions, and since we match them separately for each pattern side, we need to handle input values with more care. This is because we cannot compare them right away (we match each side of each pattern separately). Therefore, we need to register all input values using synchronisation maps and check that the mapping is correct after both sides match successfully.

Consider the example below. We can easily match both sides of the pattern, but that by itself is not enough. We also need to ensure that the input parameters `%struct.x* %0` and `i32 %1` refer to the same values in both modules.

```llvm
define i1 @diffkemp.old(%struct.x*, i32) {
    %3 = icmp ne i32 %1, 0
    ret i1 %3
}

define i1 @diffkemp.new(%struct.x*, i32) {
    call void @fun(%struct.x* %0)
    %3 = icmp ne i32 %1, 0
    ret i1 %3
}
```

I will add this example to the code as a comment as well.

Reply:
This will help a lot! Perhaps the example could be even more detailed since this function compares a *module value* with a *pattern value* while you say that in the example, we need to make sure that input parameters refer to the same values in modules (hence, I'd expect comparison of module values).

Reply:
I have actually added this example to the place where this gets called from. Having a direct example here might be quite hard to do.